src/slist.c

```diff
@@ -1 +1,332 @@
+            list_head->next = list_head->next->next;
+        }
+    }
+    return result;
+}
+
+PSINGLYLINKEDLIST_ENTRY slist_find(PSINGLYLINKEDLIST_ENTRY list_head, SLIST_MATCH_FUNCTION match_function, const void* match_context)
+{
+    PSINGLYLINKEDLIST_ENTRY result;
+    if (
+    /* Codes_SRS_SLIST_07_018: [ If list_head is NULL, slist_find shall fail and return NULL. ]*/
+    (list_head == NULL) ||
+    /* Codes_SRS_SLIST_07_019: [ If match_function is NULL, slist_find shall fail and return NULL. ]*/
+    (match_function == NULL))
+    {
+        LogError("Invalid argument (list_head=%p, match_function=%p)", list_head, match_function);
+        result = NULL;
+    }
+    else
+    {
+        PSINGLYLINKEDLIST_ENTRY current_item = list_head->next;
+
+        if (current_item == NULL)
+        {
+            /* Codes_SRS_SLIST_07_025: [ If the list is empty, slist_find shall return NULL. ]*/
+            result = NULL;
+        }
+        else
+        {
+            /* Codes_SRS_SLIST_07_020: [ slist_find shall iterate through all entris in the list and return the first entry that satisfies the match_function on success. ]*/
+            while (current_item != NULL)
+            {
+                /* Codes_SRS_SLIST_07_021: [ slist_find shall determine whether an item satisfies the match criteria by invoking the match_function for each entry in the list until a matching entry is found. ]*/
+                /* Codes_SRS_SLIST_07_022: [ The match_function shall get as arguments the list entry being attempted to be matched and the match_context as is. ]*/
+                if (match_function(current_item, match_context) == true)
+                {
+                    /* Codes_SRS_SLIST_07_024: [ If the match_function returns true, slist_find shall consider that item as matching. ]*/
+                    break;
+                }
+
+                /* Codes_SRS_SLIST_07_023: [ If the match_function returns false, slist_find shall consider that item as not matching. ]*/
+                current_item = current_item->next;
+            }
+            if (current_item == NULL)
+            {
+                /* Codes_SRS_SLIST_07_039: [ If the item is not found, slist_find shall return NULL. ]*/
+                result = NULL;
+            }
+            else
+            {
+                result = current_item;
+            }
+        }
+
+    }
+    return result;
+}
+
+int slist_remove_if(PSINGLYLINKEDLIST_ENTRY list_head, SLIST_CONDITION_FUNCTION condition_function, const void* match_context)
+{
+    int result;
+    if (
+    /* Codes_SRS_SLIST_07_026: [ If list_head is NULL, slist_remove_if shall fail and return a non-zero value. ]*/
+    (list_head == NULL) ||
+    /* Codes_SRS_SLIST_07_027: [ If condition_function is NULL, slist_remove_if shall fail and return a non-zero value. ]*/
+        (condition_function == NULL))
+    {
+        LogError("Invalid argument (list_head=%p, condition_function=%p)", list_head, condition_function);
```

Comment:
>argument [](http://example.com/codeflow?start=26&length=8)

arguments. As in plural. Probably easier to always just type that :-)

Reply:
done